lib/CodeGen/CGDynamicCheck.cpp

```diff
@@ -139,78 +142,90 @@ void CodeGenFunction::EmitDynamicBoundsCheck(const Address BaseAddr,
   const RangeBoundsExpr *SubRange = dyn_cast<RangeBoundsExpr>(SubExprBounds);
   const RangeBoundsExpr *CastRange = dyn_cast<RangeBoundsExpr>(CastBounds);
 
-  ++NumDynamicChecksRange;
+  ++NumDynamicChecksCast;
 
-  // Dynamic_check(Base == NULL || (Lower <= CastLower && CastUpper <= Upper))
-  // Emit code blocks as follows:
-  // if (Base == NULL) {...}
-  // else {
-  // if (Lower <= CastLower && CastUpper <= Upper) {...}
-  // else { trap(); llvm_unreachable(); }
-  // }
+  // Emits code as follows:
+  // %entry:
+  //   ...
+  //   %is_null = %base == null
+  //   br i1 %is_null, %success, %subsumption
+  // %subsumption:
+  //   %subsumes = (%lower <= %cast_lower && %cast_upper <= %upper)
+  //   br i1 %subsumes, %success, %failure
+  // %success:
+  //   ...
+  // %failure:
+  //   trap()
 
-  Value *Cond1 =
-            Builder.CreateIsNull(BaseAddr.getPointer(), "_Dynamic_check.null");
+  Value *IsNull =
+      Builder.CreateIsNull(BaseAddr.getPointer(), "_Dynamic_check.is_null");
 
   // Constant Folding:
-  // If we have generated a constant condition, and the condition is true,
-  // then the check will always pass and we can elide it.
-  if (const ConstantInt *ConditionConstant = dyn_cast<ConstantInt>(Cond1)) {
-    if (ConditionConstant->isOne()) {
+  // If IsNull is true (one), then a) we don't need to insert the rest
+  // of the check, computation will continue with success.
+  if (const ConstantInt *IsNullConstant = dyn_cast<ConstantInt>(IsNull)) {
+    if (IsNullConstant->isOne()) {
       ++NumDynamicChecksElided;
       return;
     }
   }
 
-  ++NumDynamicChecksInserted;
+  BasicBlock *DyCkSubsumption = createBasicBlock("_Dynamic_check.subsumption");
+  BasicBlock *DyCkSuccess = createBasicBlock("_Dynamic_check.success");
+  BasicBlock *DyCkFail = EmitDynamicCheckFailedBlock();
+
+  // Insert the IsNull Branch
+  Builder.CreateCondBr(IsNull, DyCkSuccess, DyCkSubsumption);
 
-  BasicBlock *Begin, *DyCkSuccess, *DyCkFail, *DyCkFallThrough;
-  Begin = Builder.GetInsertBlock();
-  DyCkSuccess = createBasicBlock("_Dynamic_check.succeeded");
-  DyCkFallThrough = createBasicBlock("_Dynamic_check.fallthrough", this->CurFn);
-  DyCkFail = createBasicBlock("_Dynamic_check.failed", this->CurFn);
+  // This ensures the subsumption block comes directly after the is_null branch
+  EmitBlock(DyCkSubsumption);
+
+  Builder.SetInsertPoint(DyCkSubsumption);
 
-  Builder.SetInsertPoint(DyCkFallThrough);
   // SubRange - bounds(lb, ub) vs CastRange - bounds(castlb, castub)
   // Dynamic_check(lb <= castlb && castub <= ub)
 
-  // Emit the code to generate the pointer values
+  // Emit the code to generate lb and ub
```

Comment:
`lb` and `ub` reference the two lines above, where we have `SubRange - bounds(lb, ub)`.